### PR TITLE
Check for locks when updating

### LINKED ISSUE #4: Autohupr should support auto-updating to a target release instead of latest
This block should have an environment variable that will contain the target OS version. When the target version is greater (or different) than the running version.

The user will be able to trigger an update by modifying the variable, either in a fleet or in a device.

### LINKED ISSUE #9: Autohupr should check for update locks as to not interrupt any application
Some users are running processes in the devices that can't be interrupted when the HUP process is initiated from the dashboard. HUP's current implementation doesn't have into account the containers' locks, so the HUP process will just run when invoked, killing any running process.

This interruption has resulted in some users complaining that they would have to manually control HUP when they consider it should be done by balena. 

If there's a lock in place, no update will be triggered. If there are no locks, the app will create a lock (to avoid a release update) and will then start the HUP script.

### PR DESCRIPTION
**Pull request by jaomaloy**
This resolves #9

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import { getSdk, OptionalNavigationResource } from 'balena-sdk';
 import ms, { StringValue } from 'ms';
+const lockfile = require('proper-lockfile');
```
Comment (klutchell) [inline src/main.ts:3]: This module creates directories for locks, which was not supported by supervisor until very recently, so we should use the lockfile module instead. See this commit: https://github.com/balenablocks/lockr-block/commit/f3cc33258f5c2ed22dce2fc1b51d9a1e59b46e02

**Review by klutchell — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 2]
```diff
@@ -62,9 +68,15 @@ const getTargetVersion = async (
 ): Promise<string | null> => {
 	return await balena.models.os
 		.getSupportedOsUpdateVersions(deviceType, deviceVersion)
-		.then((versions) => {
-			if (versions.recommended) {
-				return versions.recommended;
+		.then((osUpdateVersions) => {
+			if (userTargetVersion == '') {
```
Comment (klutchell) [inline src/main.ts:72]: Should we change the default behaviour to do nothing if this value is not set? We could use recommended or latest if they provide `recommended` or `latest` in the var?

**Comment by klutchell** [comment]
I think we also discussed requesting a lock while the update was in progress? Is that still planned?

File: src/main.ts [hunk 3]
```diff
@@ -82,6 +94,21 @@ const getUpdateStatus = async (uuid: string): Promise<any> => {
 	}
 };
 
+const checkLock = async (): Promise<boolean> => {
```
Comment (klutchell) [inline src/main.ts:97]: This is only checking for locks in this service container, which will never exist. We need to check for locks in any service somehow. Specifically the user application service that is already creating locks. The supervisor does this, there are probably a few ways to solve this from inside a service.

File: src/main.ts (hunk 2 above)
Comment (jaomaloy) [reply src/main.ts:72]: yeah, SGTM

File: src/main.ts (hunk 3 above)
Comment (jaomaloy) [reply src/main.ts:97]: The supervisor uses a separate lockfile binary. We could use the same.

File: src/main.ts [hunk 4]
```diff
@@ -82,6 +94,21 @@ const getUpdateStatus = async (uuid: string): Promise<any> => {
...
+				return true;
+			} else {
+				return false;
+			}
+		})
+		.catch((e: any) => {
+			console.error(e.message);
+			return false;
+		});
+};
```
Comment (Page-) [inline src/main.ts:110]: This should either actually use async/await
```suggestion
const checkLock = async (): Promise<boolean> => {
	try {
		const isLocked = await lockfile.check(lockFile)
		if (isLocked) {
			return true;
		} else {
			return false;
		}
	} catch(e: any) {
		console.error(e.message);
		return false;
	}
};
```
or not declare itself as async
```suggestion
const checkLock = (): Promise<boolean> => {
	return lockfile.check(lockFile)
		.then((isLocked: boolean) => {
			if (isLocked) {
				return true;
			} else {
				return false;
			}
		})
		.catch((e: any) => {
			console.error(e.message);
			return false;
		});
};
```

Fwiw I'd say actually using async/await would be the way to go as it has many benefits

**Comment by jaomaloy** [comment]
Closing. The HUP process checks for locks internally already. [HUP will not reboot if locks are present](https://github.com/balena-os/meta-balena/pull/3213).

**Pull request closed without merging**